### PR TITLE
Add TO Client API for Cachegroup Parameters Automation 

### PR DESCRIPTION
**Pull request by dmohan001c**
## What does this PR (Pull Request) do?

- [x] This PR is not related to any Issue

## Which Traffic Control components are affected by this PR?

- CDN in a Box
- Traffic Control Client 
- Traffic Ops
- CI tests

## What is the best way to verify this PR?
Execute all the Integration tests and make sure the tests passed.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 1]
```diff
@@ -79,6 +90,114 @@ func CreateTestCacheGroupParameters(t *testing.T) {
 	testData.CacheGroupParameterRequests = append(testData.CacheGroupParameterRequests, resp.Response...)
 }
 
+func CreateTestCacheGroupParametersAlreadyExist(t *testing.T) {
+
+	resp, _, err := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	cachegroupParameters := resp.Response
+	getAllcgparameters := cachegroupParameters.CacheGroupParameters
+	parameterID := *getAllcgparameters[0].Parameter
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:98]: There are two possible segfaults on this line, `*` and `[0]`

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 2]
```diff
@@ -79,6 +90,114 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+
+	opts := client.NewRequestOptions()
+	opts.QueryParameters.Set("name", *getAllcgparameters[0].CacheGroup)
+
+	cgResponse, reqInf, err := TOSession.GetCacheGroups(opts)
+	cg := cgResponse.Response[0]
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:104]: this will segfault if Traffic Ops returns an empty response

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 3]
```diff
@@ -79,6 +90,114 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("cannot get Parameter '%s': %v - alerts: %+v", secondParameter.Name, err, paramResp2.Alerts)
+	}
+	if len(paramResp2.Response) < 1 {
+		t.Fatalf("Expected exactly one Parameter named '%s' to exist, but found %d", secondParameter.Name, len(paramResp2.Response))
+	}
+
+	// Assign Parameter to Cache Group
+	cacheGroupID1 := cacheGroupResp1.Response[0].ID
+	cacheGroupID2 := cacheGroupResp2.Response[0].ID
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:168]: these are later dereferenced but are never checked for `nil`

File: traffic_ops/v4-client/cachegroup_parameters.go [hunk 4]
```diff
@@ -58,3 +58,10 @@ func (to *Session) CreateCacheGroupParameter(cacheGroupID, parameterID int, opts
 	reqInf, err := to.post(apiCachegroupParameters, opts, cacheGroupParameterReq, &data)
 	return data, reqInf, err
 }
+
+// CreateCacheGroupParameter associates a Parameter with a Cache Group.
```
Comment (ocket8888) [inline traffic_ops/v4-client/cachegroup_parameters.go:62]: I believe this actually associates multiple Cache Groups with multiple Parameters

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 5]
```diff
@@ -79,6 +90,114 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+
+	cgResponse, reqInf, err := TOSession.GetCacheGroups(opts)
+	cg := cgResponse.Response[0]
+	cacheGroupID := cg.ID
+	cgparameters, reqInf, err := TOSession.CreateCacheGroupParameter(*cacheGroupID, parameterID, client.RequestOptions{})
+	if reqInf.StatusCode != http.StatusBadRequest {
+		t.Errorf("Expected 400 status code, got %v", reqInf.StatusCode)
+	}
+	if err == nil {
+		t.Errorf("Expected Parameter already associated with cachegroup %v - Alerts %v", cgparameters, cgparameters.Alerts)
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:111]: I think you meant to print the Cache Group ID here instead of `cgparameters`

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 6]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+
+func SortTestCacheGroupParameters(t *testing.T) {
+	resp, _, err := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	if err != nil {
+		t.Fatalf("Expected no error, but got: %v - alerts: %+v", err, resp.Alerts)
+	}
+	cachegroupParameters := resp.Response
+	sortedList := make([]string, 0, len(cachegroupParameters.CacheGroupParameters))
+	for _, cgparameters := range cachegroupParameters.CacheGroupParameters {
+		sortedList = append(sortedList, strconv.Itoa(*cgparameters.Parameter))
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:356]: this will segfault if the parameter was null or undefined in the response

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 7]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+		t.Fatalf("Traffic Ops returned %d CacheGroup Parameters using default sort order, but %d CacheGroup Parameters when sort order was explicitly set to descending", len(respAsc), len(respDesc))
+	}
+
+	// reverse the descending-sorted response and compare it to the ascending-sorted one
+	// TODO ensure at least two in each slice? A list of length one is
+	// trivially sorted both ascending and descending.
+	for start, end := 0, len(respDesc)-1; start < end; start, end = start+1, end-1 {
+		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
+	}
+	if *respDesc[0].Parameter != *respAsc[0].Parameter {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:400]: This will segfault if either Parameter ID was null or undefined in the response

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 8]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+		t.Errorf("CacheGroup Parameters responses are not equal after reversal: Asc: %v - Desc: %v", *respDesc[0].Parameter, *respAsc[0].Parameter)
+	}
+}
+
+func CreateTestCacheGroupParametersWithInvalidData(t *testing.T) {
+	resp, _, err := TOSession.GetParameters(client.RequestOptions{})
+	parametersResp := resp.Response[0]
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:407]: this will segfault if the response was null, undefined, or empty

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 9]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	for start, end := 0, len(respDesc)-1; start < end; start, end = start+1, end-1 {
+		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
+	}
+	if *respDesc[0].Parameter != *respAsc[0].Parameter {
+		t.Errorf("CacheGroup Parameters responses are not equal after reversal: Asc: %v - Desc: %v", *respDesc[0].Parameter, *respAsc[0].Parameter)
+	}
+}
+
+func CreateTestCacheGroupParametersWithInvalidData(t *testing.T) {
+	resp, _, err := TOSession.GetParameters(client.RequestOptions{})
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:406]: The value of this error should be checked if it's being stored (really I think they should just always check it)

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 10]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	cgparameters, reqInf, err := TOSession.CreateCacheGroupParameter(0, parameterID, client.RequestOptions{})
+	if reqInf.StatusCode != http.StatusNotFound {
+		t.Errorf("Expected 404 status code, got %v", reqInf.StatusCode)
+	}
+	if err == nil {
+		t.Errorf("Expected cachegroup not found, but found Alerts %v", cgparameters.Alerts)
+	}
+
+	cgresp, _, err := TOSession.GetCacheGroups(client.RequestOptions{})
+	cachegroupResp := cgresp.Response[0]
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:419]: This will segfault if the response was null, undefined, or empty

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 11]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	cachegroupID := cachegroupResp.ID
+
+	//Create CacheGroupParameters with Invalid Parameter ID
+	cgparametersresp, reqInf, err := TOSession.CreateCacheGroupParameter(*cachegroupID, 0, client.RequestOptions{})
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:423]: This will segfault if the Cache Group ID was null or undefined in the response

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 12]
```diff
@@ -155,3 +274,201 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	opts.QueryParameters.Set("parameter", "0")
+	resp, _, _ = TOSession.GetAllCacheGroupParameters(opts)
+	cachegroupparameters = resp.Response
+	if len(cachegroupparameters.CacheGroupParameters) > 0 {
+		t.Errorf("Expected No response for the invalid parameter id, but found %d responses", len(cachegroupparameters.CacheGroupParameters))
+	}
+}
+
+func GetTestCacheGroupParameter(t *testing.T) {
+	firstData := testData.CacheGroupParameterRequests[0]
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:453]: This will segfault if the test data is in a bad state

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 1 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:98]: added null checks.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 5 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:111]: updated the error statement.

File: traffic_ops/v4-client/cachegroup_parameters.go (hunk 4 above)
Comment (dmohan001c) [reply traffic_ops/v4-client/cachegroup_parameters.go:62]: updated the comments.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 3 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:168]: already evaluated nil checks at the above line. if len(cacheGroupResp2.Response) != 1

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 6 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:356]: Added null checks

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 7 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:400]: Added null checks before this comparison

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 8 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:407]: added null checks.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 9 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:406]: err checking has added.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 10 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:419]: added null checks

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 11 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:423]: added null checks

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 12 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:453]: Fixed it.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 3 above)
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:178]: Yes, `cacheGroupResp1.Response` can't be `nil`, and `cacheGroupResp1.Response[0]` won't segfault. These lines won't segfault at all. But `cacheGroupID1` and `cacheGroupID2` are pointer values that themselves could be `nil`, and they are later dereferenced without checking for that.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 1 above)
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:101]: This can still segfault if the response did not contain a `parameter` property at index `0`

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 2 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go]: Fixed.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 13]
```diff
@@ -79,6 +90,124 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
 
+func CreateTestCacheGroupParametersAlreadyExist(t *testing.T) {
+
+	resp, _, _ := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	cachegroupParameters := resp.Response
+	getAllcgparameters := cachegroupParameters.CacheGroupParameters
+	if len(getAllcgparameters) < 1 {
+		t.Fatalf("No existing cachegroup parameters available to test duplicate functionality")
+	}
+	parameterID := *getAllcgparameters[0].Parameter
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:101]: This line will segfault if the response improperly returned a `null` or undefined `parameter` property in the first element of the response array

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 14]
```diff
@@ -79,6 +90,124 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+
+	//Get cachegroup id from cachegroup name
+	opts := client.NewRequestOptions()
+	opts.QueryParameters.Set("name", *getAllcgparameters[0].CacheGroup)
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:105]: This line will segfault if the response improperly returned a `null` or undefined `cacheGroup` ('cachegroup`? `cache_group`?) property in the first element of the response array

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 15]
```diff
@@ -79,6 +90,124 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+
+	cgResponse, _, _ := TOSession.GetCacheGroups(opts)
+	if len(cgResponse.Response) < 1 {
+		t.Fatalf("No data available for GetCacheGroups by Name - Length %d", len(cgResponse.Response))
+	}
+	cacheGroupID := cgResponse.Response[0].ID
+	if parameterID != 0 && *cacheGroupID != 0 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:112]: This line will segfault if the response improperly returned a `null` or undefined `id` property in the first element of the response array

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 16]
```diff
@@ -79,6 +90,124 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+		cgparameters, reqInf, err := TOSession.CreateCacheGroupParameter(*cacheGroupID, parameterID, client.RequestOptions{})
+		if reqInf.StatusCode != http.StatusBadRequest {
+			t.Errorf("Expected 400 status code, got %v", reqInf.StatusCode)
+		}
+		if err == nil {
+			t.Errorf("Expected Parameter - %d already associated with cachegroup %d - Alerts %v", parameterID, cacheGroupID, cgparameters.Alerts)
+		}
+	} else {
+		t.Errorf("Nil value found in one of the values. Cachegroup ID %d, Parameter ID %d", cacheGroupID, parameterID)
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:121]: This will print the pointer value of `cacheGroupID`, not the actual ID value to which it points.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 17]
```diff
@@ -155,3 +284,226 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	sortedList := make([]string, 0, len(cachegroupParameters.CacheGroupParameters))
+	for _, cgparameters := range cachegroupParameters.CacheGroupParameters {
+		if cgparameters.Parameter != nil {
+			sortedList = append(sortedList, strconv.Itoa(*cgparameters.Parameter))
+		}
+	}
+
+	res := sort.SliceIsSorted(sortedList, func(p, q int) bool {
+		return sortedList[p] < sortedList[q]
+	})
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:373]: for what it's worth, `sort` provides [`sort.StringsAreSorted`](https://pkg.go.dev/sort#StringsAreSorted) which does exactly what this function does in one line.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 18]
```diff
@@ -155,3 +284,226 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+		t.Fatalf("Traffic Ops returned %d CacheGroup Parameters using default sort order, but %d CacheGroup Parameters when sort order was explicitly set to descending", len(respAsc), len(respDesc))
+	}
+
+	// reverse the descending-sorted response and compare it to the ascending-sorted one
+	// TODO ensure at least two in each slice? A list of length one is
+	// trivially sorted both ascending and descending.
+	for start, end := 0, len(respDesc)-1; start < end; start, end = start+1, end-1 {
+		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
+	}
+	if *respDesc[0].Parameter != 0 && *respAsc[0].Parameter != 0 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:412]: This line will segfault if either of the responses' first response element had a `null` or undefined `parameter` property

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 19]
```diff
@@ -155,3 +284,226 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+	}
+	cgresp, _, err := TOSession.GetCacheGroups(client.RequestOptions{})
+	if len(cgresp.Response) < 1 {
+		t.Fatal("Need at least one cachegroups to test invalid Cache Groups functionality")
+	}
+	if err != nil {
+		t.Fatalf("Error in fetching get Cachegroups %v", err)
+	}
+
+	if parameterresp.Response[0].ID != 0 && *cgresp.Response[0].ID != 0 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:437]: This line will segfault if `cgresp.Response[0].ID` is `nil`

File: traffic_ops/testing/api/v4/divisions_test.go [hunk 20]
```diff
@@ -233,8 +233,12 @@ func SortTestDivisionDesc(t *testing.T) {
...
 		respDesc.Response[start], respDesc.Response[end] = respDesc.Response[end], respDesc.Response[start]
 	}
-	if respDesc.Response[0].Name != respAsc.Response[0].Name {
-		t.Errorf("Division responses are not equal after reversal: %s - %s", respDesc.Response[0].Name, respAsc.Response[0].Name)
+	if len(respDesc.Response[0].Name) > 0 && len(respAsc.Response[0].Name) > 0 {
+		if respDesc.Response[0].Name != respAsc.Response[0].Name {
+			t.Errorf("Division responses are not equal after reversal: %s - %s", respDesc.Response[0].Name, respAsc.Response[0].Name)
+		}
+	} else {
+		t.Errorf("Division name shouldn't be empty while sorting the response")
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/divisions_test.go:241]: This change seems unrelated to this PR

File: traffic_ops/testing/api/v4/coordinates_test.go [hunk 21]
```diff
@@ -212,8 +212,12 @@ func SortTestCoordinatesDesc(t *testing.T) {
...
 		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
 	}
-	if respDesc[0].Name != respAsc[0].Name {
-		t.Errorf("Coordinates responses are not equal after reversal: %s - %s", respDesc[0].Name, respAsc[0].Name)
+	if len(respDesc[0].Name) > 0 && len(respAsc[0].Name) > 0 {
+		if respDesc[0].Name != respAsc[0].Name {
+			t.Errorf("Coordinates responses are not equal after reversal: %s - %s", respDesc[0].Name, respAsc[0].Name)
+		}
+	} else {
+		t.Errorf("Coordinates name shouldn't be empty while sorting the response")
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/coordinates_test.go:220]: This change seems unrelated to this PR

File: traffic_ops/testing/api/v4/regions_test.go [hunk 22]
```diff
@@ -210,8 +210,12 @@ func SortTestRegionsDesc(t *testing.T) {
...
 		respDesc[start], respDesc[end] = respDesc[end], respDesc[start]
 	}
-	if respDesc[0].Name != respAsc[0].Name {
-		t.Errorf("Regions responses are not equal after reversal: Asc: %s - Desc: %s", respDesc[0].Name, respAsc[0].Name)
+	if len(respDesc[0].Name) > 0 && len(respAsc[0].Name) > 0 {
+		if respDesc[0].Name != respAsc[0].Name {
+			t.Errorf("Regions responses are not equal after reversal: Asc: %s - Desc: %s", respDesc[0].Name, respAsc[0].Name)
+		}
+	} else {
+		t.Errorf("Region name shouldn't be empty while sorting the response")
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/regions_test.go:218]: This change seems unrelated to this PR

File: traffic_ops/v4-client/cachegroup_parameters.go [hunk 23]
```diff
@@ -58,3 +58,10 @@ func (to *Session) CreateCacheGroupParameter(cacheGroupID, parameterID int, opts
 	reqInf, err := to.post(apiCachegroupParameters, opts, cacheGroupParameterReq, &data)
 	return data, reqInf, err
 }
+
+// CreateMultipleCacheGroupParameter associates a multiple parameter with a multiple Cache Group.
```
Comment (ocket8888) [inline traffic_ops/v4-client/cachegroup_parameters.go:62]: > ... a multiple parameter with a multiple Cache Group.

nit: grammar

should be

> ... associates multiple Parameters with multiple Cache Groups.

**Review by ocket8888 — CHANGES_REQUESTED** [review]

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 13 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:101]: Fixed null checks

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 14 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:105]: Fixed.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 15 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:112]: Fixed.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 16 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:121]: Fixed it.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 17 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:373]: Fixed.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 18 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:412]: Added null checks.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 19 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:440]: Added null checks

File: traffic_ops/testing/api/v4/divisions_test.go (hunk 20 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/divisions_test.go:241]: Just added null checks, and PR related to theis functionality has closed. Do you want me to include these changes in the respective PR?

File: traffic_ops/testing/api/v4/coordinates_test.go (hunk 21 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/coordinates_test.go:220]: Just added null checks, and PR related to theis functionality has closed. Do you want me to include these changes in the respective PR?

File: traffic_ops/testing/api/v4/regions_test.go (hunk 22 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/regions_test.go:218]: Just added null checks, and PR related to theis functionality has closed. Do you want me to include these changes in the respective PR?

File: traffic_ops/v4-client/cachegroup_parameters.go (hunk 23 above)
Comment (dmohan001c) [reply traffic_ops/v4-client/cachegroup_parameters.go]: changed,

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 24]
```diff
@@ -79,6 +90,129 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
 
+func CreateTestCacheGroupParametersAlreadyExist(t *testing.T) {
+
+	resp, _, _ := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	cachegroupParameters := resp.Response
+	getAllcgparameters := cachegroupParameters.CacheGroupParameters
+	if len(getAllcgparameters) < 1 {
+		t.Fatalf("No existing cachegroup parameters available to test duplicate functionality")
+	}
+	if getAllcgparameters[0].Parameter == nil && *getAllcgparameters[0].Parameter != 0 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:101]: The second part of this check will always segfault if it gets checked. Also, I think a Parameter ID of 0 is always invalid, so I don't think we want to enforce that. Seems like the second part of this check just shouldn't be there.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 25]
```diff
@@ -79,6 +90,129 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+		t.Fatalf("found nil/0 Value in Paramters")
+	}
+	if len(*getAllcgparameters[0].CacheGroup) < 1 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:104]: This line will segfault if Traffic Ops incorrectly responded with a null or undefined Cache Group - also, I'm not sure why it's checking the length of that string. Yes an empty name is invalid, but it won't cause a segfault and will be caught later because the client call on line 113 will return an empty array

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 26]
```diff
@@ -79,6 +90,129 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+
+	//Get cachegroup id from cachegroup name
+	opts := client.NewRequestOptions()
+	opts.QueryParameters.Set("name", *getAllcgparameters[0].CacheGroup)
+
+	cgResponse, _, _ := TOSession.GetCacheGroups(opts)
+	if len(cgResponse.Response) < 1 {
+		t.Fatalf("No data available for GetCacheGroups by Name - Length %d", len(cgResponse.Response))
+	}
+	if cgResponse.Response[0].ID == nil && *cgResponse.Response[0].ID != 0 {
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:117]: The error says that it found a nil or 0 value for the Cache Group ID, but that will only get printed if the ID is both nil AND NOT 0. Which, by the way, also means that if the second half of this condition is ever checked, it will always segfault.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 27]
```diff
@@ -79,6 +90,129 @@ func CreateTestCacheGroupParameters(t *testing.T) {
...
+	firstParameter := testData.Parameters[1]
+	secondParameter := testData.Parameters[2]
+	opts.QueryParameters.Set("name", firstParameter.Name)
+	paramResp1, _, err := TOSession.GetParameters(opts)
+	if err != nil {
+		t.Errorf("cannot get Parameter '%s': %v - alerts: %+v", firstParameter.Name, err, paramResp1.Alerts)
+	}
+	if len(paramResp1.Response) != 1 {
+		t.Fatalf("Expected exactly one Parameter named '%s' to exist, but found %d", firstParameter.Name, len(paramResp1.Response))
+	}
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:171]: Names don't uniquely identify Parameters. In fact, the second two Parameters in the testing data could easily have the same Name, they just coincidentally don't, currently. So you can't ever really expect only one Parameter to exist with a given Name without any other constraints.

File: traffic_ops/v4-client/cachegroup_parameters.go [hunk 28]
```diff
@@ -58,3 +58,10 @@ func (to *Session) CreateCacheGroupParameter(cacheGroupID, parameterID int, opts
 	reqInf, err := to.post(apiCachegroupParameters, opts, cacheGroupParameterReq, &data)
 	return data, reqInf, err
 }
+
+// CreateMultipleCacheGroupParameter associates multiple parameter with multiple Cache Group.
```
Comment (ocket8888) [inline traffic_ops/v4-client/cachegroup_parameters.go:62]: nit: "parameter" and "Cache Group" should be plural: "Parameters" and "Cache Groups"

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go [hunk 29]
```diff
@@ -155,3 +289,224 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
...
+func SortTestCacheGroupParameters(t *testing.T) {
+	resp, _, err := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	if err != nil {
+		t.Fatalf("Expected no error, but got: %v - alerts: %+v", err, resp.Alerts)
+	}
+	cachegroupParameters := resp.Response
+	sortedList := make([]string, 0, len(cachegroupParameters.CacheGroupParameters))
+	for _, cgparameters := range cachegroupParameters.CacheGroupParameters {
+		if cgparameters.Parameter != nil {
+			sortedList = append(sortedList, strconv.Itoa(*cgparameters.Parameter))
```
Comment (ocket8888) [inline traffic_ops/testing/api/v4/cachegroups_parameters_test.go:372]: This line will segfault if Traffic Ops returned an improper response with null or undefined `parameter` property in any element.

File: traffic_ops/testing/api/v4/coordinates_test.go (hunk 21 above)
Comment (ocket8888) [reply traffic_ops/testing/api/v4/coordinates_test.go:220]: yes, that would help keep this one cleaner

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 27 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:171]: Fixed it.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 29 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:372]: I hope this line will take care of the null check. 
if cgparameters.Parameter != nil{

File: traffic_ops/v4-client/cachegroup_parameters.go (hunk 28 above)
Comment (dmohan001c) [reply traffic_ops/v4-client/cachegroup_parameters.go:62]: Fixed.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 26 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:117]: Fixed it.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 25 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:104]: Fixed it.

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 24 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:101]: removed the second part of the validation.

File: traffic_ops/testing/api/v4/coordinates_test.go (hunk 21 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/coordinates_test.go:220]: Okay.

File: traffic_ops/testing/api/v4/divisions_test.go (hunk 20 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/divisions_test.go:241]: removed this changes.

File: traffic_ops/testing/api/v4/coordinates_test.go (hunk 21 above)
Comment (dmohan001c) [reply traffic_ops/testing/api/v4/coordinates_test.go:220]: removed this changes

File: traffic_ops/testing/api/v4/cachegroups_parameters_test.go (hunk 29 above)
Comment (ocket8888) [reply traffic_ops/testing/api/v4/cachegroups_parameters_test.go:372]: You're right, I missed that, somehow

**Review by ocket8888 — APPROVED** [review]
